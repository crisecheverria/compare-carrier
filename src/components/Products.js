import React, { useState, Fragment } from "react"
import PropTypes from "prop-types"
import getProducts from "../services/fakeProductService"
import splitEvery from "../utils"
import Product from "./Product"

function Products({ sortParameter }) {
  const [products] = useState(getProducts().items)

  function compareCheapest(a, b) {
    if (a.cost > b.cost) return 1
    if (a.cost < b.cost) return -1
    return 0
  }

  function compareFastest(a, b) {
    if (a.lead_time > b.lead_time) return 1
    if (a.lead_time < b.lead_time) return -1
    return 0
  }

  const sortedProducts = [...products]
  if (sortParameter === "cheapest") sortedProducts.sort(compareCheapest)

  if (sortParameter === "fastest") sortedProducts.sort(compareFastest)

  return (
    <Fragment>
      {splitEvery(sortedProducts, 3).map((productsChunk) => (
        <div className="columns is-desktop">
          {productsChunk.map((product) => (
            <div className="column">
              <Product key={product.product} product={product} />
            </div>
          ))}
        </div>
      ))}
    </Fragment>
  )
}

Products.propTypes = {
  sortParameter: PropTypes.string,
}

export default Products
