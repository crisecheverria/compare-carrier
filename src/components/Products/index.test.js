import React from "react"
import { render } from "@testing-library/react"
import Products from "."
import getProducts from "../../services/fakeProductService"

test("render <Products /> component", () => {
  const sortParameter = "cheapest"
  const { baseElement } = render(<Products sortParameter={sortParameter} />)

  expect(baseElement).toMatchSnapshot()
})

test("render number of carriers from 'products.json' component", () => {
  const sortParameter = "cheapest"
  const { getAllByTestId } = render(<Products sortParameter={sortParameter} />)
  const products = getAllByTestId(/products/i)

  expect(products.length).toBe(getProducts().items.length)
})
