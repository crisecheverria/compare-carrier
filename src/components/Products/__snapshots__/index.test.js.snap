// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render <Products /> component 1`] = `
<body>
  <div>
    <div
      class="columns is-desktop"
    >
      <div
        class="column"
        data-testid="products"
      >
        <div
          class="custom-box"
        >
          <article
            class="media"
          >
            <div
              class="media-left"
            >
              <figure
                class="image is-64x64"
              >
                <img
                  alt="DHL 1-logo"
                  class="is-rounded"
                  src="images/dhl.svg"
                />
              </figure>
            </div>
            <div
              class="media-content"
            >
              <div
                class="content"
              >
                <p
                  data-testid="carrier"
                >
                  DHL 1
                </p>
                <p
                  class="currency"
                  data-testid="cost"
                >
                  <strong>
                    62.98
                  </strong>
                </p>
                <span>
                  <strong
                    data-testid="time"
                  >
                    2 days
                  </strong>
                  <progress
                    class="progress is-warning"
                    max="7"
                    value="2"
                  />
                </span>
              </div>
            </div>
          </article>
        </div>
      </div>
      <div
        class="column"
        data-testid="products"
      >
        <div
          class="custom-box"
        >
          <article
            class="media"
          >
            <div
              class="media-left"
            >
              <figure
                class="image is-64x64"
              >
                <img
                  alt="TNT 2-logo"
                  class="is-rounded"
                  src="images/tnt.svg"
                />
              </figure>
            </div>
            <div
              class="media-content"
            >
              <div
                class="content"
              >
                <p
                  data-testid="carrier"
                >
                  TNT 2
                </p>
                <p
                  class="currency"
                  data-testid="cost"
                >
                  <strong>
                    98.23
                  </strong>
                </p>
                <span>
                  <strong
                    data-testid="time"
                  >
                    3 days
                  </strong>
                  <progress
                    class="progress is-warning"
                    max="7"
                    value="3"
                  />
                </span>
              </div>
            </div>
          </article>
        </div>
      </div>
      <div
        class="column"
        data-testid="products"
      >
        <div
          class="custom-box"
        >
          <article
            class="media"
          >
            <div
              class="media-left"
            >
              <figure
                class="image is-64x64"
              >
                <img
                  alt="UPS 2-logo"
                  class="is-rounded"
                  src="images/ups.svg"
                />
              </figure>
            </div>
            <div
              class="media-content"
            >
              <div
                class="content"
              >
                <p
                  data-testid="carrier"
                >
                  UPS 2
                </p>
                <p
                  class="currency"
                  data-testid="cost"
                >
                  <strong>
                    123.00
                  </strong>
                </p>
                <span>
                  <strong
                    data-testid="time"
                  >
                    2 days
                  </strong>
                  <progress
                    class="progress is-warning"
                    max="7"
                    value="2"
                  />
                </span>
              </div>
            </div>
          </article>
        </div>
      </div>
    </div>
    <div
      class="columns is-desktop"
    >
      <div
        class="column"
        data-testid="products"
      >
        <div
          class="custom-box"
        >
          <article
            class="media"
          >
            <div
              class="media-left"
            >
              <figure
                class="image is-64x64"
              >
                <img
                  alt="UPS 1-logo"
                  class="is-rounded"
                  src="images/ups.svg"
                />
              </figure>
            </div>
            <div
              class="media-content"
            >
              <div
                class="content"
              >
                <p
                  data-testid="carrier"
                >
                  UPS 1
                </p>
                <p
                  class="currency"
                  data-testid="cost"
                >
                  <strong>
                    134.00
                  </strong>
                </p>
                <span>
                  <strong
                    data-testid="time"
                  >
                    2 days
                  </strong>
                  <progress
                    class="progress is-warning"
                    max="7"
                    value="2"
                  />
                </span>
              </div>
            </div>
          </article>
        </div>
      </div>
      <div
        class="column"
        data-testid="products"
      >
        <div
          class="custom-box"
        >
          <article
            class="media"
          >
            <div
              class="media-left"
            >
              <figure
                class="image is-64x64"
              >
                <img
                  alt="DHL 2-logo"
                  class="is-rounded"
                  src="images/dhl.svg"
                />
              </figure>
            </div>
            <div
              class="media-content"
            >
              <div
                class="content"
              >
                <p
                  data-testid="carrier"
                >
                  DHL 2
                </p>
                <p
                  class="currency"
                  data-testid="cost"
                >
                  <strong>
                    147.23
                  </strong>
                </p>
                <span>
                  <strong
                    data-testid="time"
                  >
                    2 days
                  </strong>
                  <progress
                    class="progress is-warning"
                    max="7"
                    value="2"
                  />
                </span>
              </div>
            </div>
          </article>
        </div>
      </div>
      <div
        class="column"
        data-testid="products"
      >
        <div
          class="custom-box"
        >
          <article
            class="media"
          >
            <div
              class="media-left"
            >
              <figure
                class="image is-64x64"
              >
                <img
                  alt="DSV 1-logo"
                  class="is-rounded"
                  src="images/dsv.svg"
                />
              </figure>
            </div>
            <div
              class="media-content"
            >
              <div
                class="content"
              >
                <p
                  data-testid="carrier"
                >
                  DSV 1
                </p>
                <p
                  class="currency"
                  data-testid="cost"
                >
                  <strong>
                    176.00
                  </strong>
                </p>
                <span>
                  <strong
                    data-testid="time"
                  >
                    2 days
                  </strong>
                  <progress
                    class="progress is-warning"
                    max="7"
                    value="2"
                  />
                </span>
              </div>
            </div>
          </article>
        </div>
      </div>
    </div>
    <div
      class="columns is-desktop"
    >
      <div
        class="column"
        data-testid="products"
      >
        <div
          class="custom-box"
        >
          <article
            class="media"
          >
            <div
              class="media-left"
            >
              <figure
                class="image is-64x64"
              >
                <img
                  alt="TNT 3-logo"
                  class="is-rounded"
                  src="images/tnt.svg"
                />
              </figure>
            </div>
            <div
              class="media-content"
            >
              <div
                class="content"
              >
                <p
                  data-testid="carrier"
                >
                  TNT 3
                </p>
                <p
                  class="currency"
                  data-testid="cost"
                >
                  <strong>
                    349.21
                  </strong>
                </p>
                <span>
                  <strong
                    data-testid="time"
                  >
                    1 day
                  </strong>
                  <progress
                    class="progress is-warning"
                    max="7"
                    value="1"
                  />
                </span>
              </div>
            </div>
          </article>
        </div>
      </div>
      <div
        class="column"
        data-testid="products"
      >
        <div
          class="custom-box"
        >
          <article
            class="media"
          >
            <div
              class="media-left"
            >
              <figure
                class="image is-64x64"
              >
                <img
                  alt="TNT 1-logo"
                  class="is-rounded"
                  src="images/tnt.svg"
                />
              </figure>
            </div>
            <div
              class="media-content"
            >
              <div
                class="content"
              >
                <p
                  data-testid="carrier"
                >
                  TNT 1
                </p>
                <p
                  class="currency"
                  data-testid="cost"
                >
                  <strong>
                    416.21
                  </strong>
                </p>
                <span>
                  <strong
                    data-testid="time"
                  >
                    1 day
                  </strong>
                  <progress
                    class="progress is-warning"
                    max="7"
                    value="1"
                  />
                </span>
              </div>
            </div>
          </article>
        </div>
      </div>
      <div
        class="column"
        data-testid="products"
      >
        <div
          class="custom-box"
        >
          <article
            class="media"
          >
            <div
              class="media-left"
            >
              <figure
                class="image is-64x64"
              >
                <img
                  alt="DHL 3-logo"
                  class="is-rounded"
                  src="images/dhl.svg"
                />
              </figure>
            </div>
            <div
              class="media-content"
            >
              <div
                class="content"
              >
                <p
                  data-testid="carrier"
                >
                  DHL 3
                </p>
                <p
                  class="currency"
                  data-testid="cost"
                >
                  <strong>
                    1342.76
                  </strong>
                </p>
                <span>
                  <strong
                    data-testid="time"
                  >
                    1 day
                  </strong>
                  <progress
                    class="progress is-warning"
                    max="7"
                    value="1"
                  />
                </span>
              </div>
            </div>
          </article>
        </div>
      </div>
    </div>
  </div>
</body>
`;
